### ISSUE TITLE
Datepicker: Incorrect date displayed when selecting a date in the first week of November 1999

### ISSUE THREAD

**Issue by willisd2** [angular/material2#5185]
#### Bug, feature request, or proposal:
Bug

#### What is the expected behavior?
When selecting November, 4, 1999 from the datepicker, the input will display 11/04/1999

#### What is the current behavior?
When selecting November, 4, 1999 from the datepicker the input displays 1 day behind, 11/03/1999

#### What are the steps to reproduce?
Select November 4, 1999 from the datepicker in the supplied plunkr:
http://plnkr.co/edit/4kiAVKRyskrJFxHNZLuy

#### What is the use-case or motivation for changing an existing behavior?
It's a bug, the existing behavior is unintended

#### Which versions of Angular, Material, OS, TypeScript, browsers are affected?
Not sure, just something I noticed on the Material documentation site.  I'm using Chrome 58.0.3029.110 as my browser 

#### Is there anything else we should know?
This happens for the first week of November in both 1999 and 2000, likely other places as well.

**Comment by julianobrasil** [comment]
Seem to be working fine:

![image](https://user-images.githubusercontent.com/25405143/27254861-f5887338-5367-11e7-9ddb-e00e1925b9ab.png)

**Comment by willisd2** [comment]
![angularmaterialdatepickerissue](https://user-images.githubusercontent.com/9123932/27254941-6996bb24-5361-11e7-8479-eea983f3b5e0.png)

Interesting, might be a locale issue.

**Comment by julianobrasil** [comment]
I forked your plunker and added some locale settings. Try it with your language (don't forget to fix the `myParse` function): http://plnkr.co/edit/xfab14?p=preview

**Comment by willisd2** [comment]
Hmm, I'm still getting the same behavior, in Chrome and Firefox.  What needs to be fixed in the myParse function?  
Also, FWIW I tried it in Microsoft Edge 38.14393.0.0 and I get the _expected_ behavior.

**Comment by julianobrasil** [comment]
The parse myParse function is configured to brazilian portuguese, which uses the `day/month/year` format. If the format of your language is similar, you have to do nothing.

Intriguing behavior you're experiencing...

**Comment by julianobrasil** [comment]
It happens here. I picked a date way before 1999.

![image](https://user-images.githubusercontent.com/25405143/27255214-224fcfb2-5370-11e7-885d-7ab5ef40b6b3.png)

**Comment by julianobrasil** [comment]
It's a problem with Daylight Saving Time. The datepicker builds a date with the date you picked and set the time to 0:0:0 (in fact, javascript does it when you build a Date object providing year, month and day).
You can get more details in here: https://stackoverflow.com/a/5860841/6433166

You can solve it overriding the createDate function of the NativeDateAdapter and setting the hours to 12, instead of leaving it to 0:0:0 (midnight):

Take a look at this plunker where I made the little change shown below: http://plnkr.co/edit/shFewF?p=preview

```
createDate(year: number, month: number, date: number): Date {
    
    if (month < 0 || month > 11 || date < 1) {
      return null;
    }

    let result = this._createDateWithOverflow(year, month, date);
    result.setHours(12); // <- here is what corrects it

    if (result.getMonth() != month) {
      return null;
    }

    return result;
  }
```

@fxck , please, take a look at this... is my conclusion right?

**Comment by mmalerba** [comment]
Interesting, I'm not seeing this for 11/4/1999 or 11/6/1984, I wonder if it only happens depending on the time zone you're in or something

**Comment by julianobrasil** [comment]
@mmalerba, I think the problem hapens when you try to access a date contained in Daylight Summer Time in your locale zone.

**Comment by julianobrasil** [comment]
The problem seems to affect some past dates, according to what is writen in [ECMA 5.1 specification](https://www.ecma-international.org/ecma-262/5.1/):

_The implementation of ECMAScript should not try to determine whether the exact time was subject to daylight saving time, but just whether daylight saving time would have been in effect if the current daylight saving time algorithm had been used at the time. This avoids complications such as taking into account the years that the locale observed daylight saving time year round._

Basically it seems that it'll calculate past DST dates as they had happened nowadays.

**Comment by mmalerba** [comment]
@julianobrasil Ah yeah that sounds like it may be the issue, I wonder if I can try converting to UTC dates or something to get around this. I'd like to avoid setting hours since it'll complicate things later when I add a timepicker

**Comment by julianobrasil** [comment]
@mmalerba, I tried it with UTC dates: http://plnkr.co/edit/6WOeQX?p=preview

It seems to work fine with the problematic 1984-11-7 I had tried before with no need to mess with the hours:

![image](https://user-images.githubusercontent.com/25405143/27517374-d0adc23a-59a1-11e7-85fe-ff199e1b2e31.png)

I also tried some other timezones (changing my notebook timezone) that are critical from my point of view (GMT:0, GMT:+1200 and GMT:-1200) during standard time and DST. It was not an exhaustive testing series, but it worked. Also tested with GMT:-0500 (as it seems to be the timezone @willisd2 had the problem)

**Comment by jimitndiaye** [comment]
I'm seeing the behavior with 10/26/1982 - menu displays correct date but value is 10/25/1982. Time zone is BST.

**Comment by julianobrasil** [comment]
@jimitndiaye, did you tested that date with my plunker? I tried here setting to UTC+01:00 in my notebook (whitch I think is BST - correct me if I'm wrong) and it worked fine with 10/26/1982. 

If you didn't try my plunker, please, give it a try (it's in Brazilian Date Format: dd/MM/yyyy) and let me know if it doesn't work: http://plnkr.co/edit/6WOeQX?p=preview

**Comment by jimitndiaye** [comment]
@julianobrasil Works fine with your plunker. I originally tried it on the demo site: https://material.angular.io/components/datepicker/examples

**Comment by craftyInstigator** [comment]
![image](https://user-images.githubusercontent.com/5348503/28219229-3b81cbc4-6889-11e7-8ee0-7834ece59801.png)

**Comment by julianobrasil** [comment]
@craftyInstigator, is has a PR to fix it. If you can't wait, try this solution: http://plnkr.co/edit/jc4Ywu?p=preview

**Comment by murilobd** [comment]
Hey @julianobrasil I'm facing the same issue right now! It's 17th October 17:54 (GMT-02 because of daylight summer time) right now and when I select a date on DatePicker (using the CustomDateAdapter from [http://plnkr.co/edit/jc4Ywu?p=preview](http://plnkr.co/edit/jc4Ywu?p=preview)) I'm getting the next day (18th October).
I've tried a bunch of other configurations, changing hours, trying with UTC but didn't manage to make it work correctly. Any idea what else I could do?

**Comment by julianobrasil** [comment]
@murilobd, aparen't my there's an issue with datepicker and DST in some regions. This seems to be related to Javascript Date object and ECMA 5.1 - so it depends on the browser you're using. Take a look at https://github.com/angular/material2/issues/7167#issuecomment-336313813.

**Comment by murilobd** [comment]
Got it @julianobrasil ! Thanks for answering!! Last question: how complicated would be to use moment.js as DateAdapter!? Maybe they already have fixed this issue!

**Comment by julianobrasil** [comment]
Yeah, in moment.js dateadapter this issue is not happening.  I've never used it, but if I had problems, I'd use for sure! It doesn't seem a complicated thing. Aparently you'll just have to switch your project Dates to Moment object.

**Comment by murilobd** [comment]
Thanks @julianobrasil !! I just created a gist with the MomentDateAdapter I've created for build beta 8 from Angular. If anyone needs: https://gist.github.com/murilobd/605edbc5c668f4f9f943dfb82a7cadd9

**Comment by julianobrasil** [comment]
@murilobd, currently there is an official MomentDateAdapter (that accidentaly wasn't included in beta.12): https://github.com/angular/material2-moment-adapter-builds

**Comment by murilobd** [comment]
@julianobrasil yeap! I found this MomentDatePicker and used! As my application is running on Beta 8, I had to change some code to make it work!

**Comment by rahimnagori** [comment]
This issue only appears if your current timezone is below 00:00 (negative).
